Zen 450 fix native apps

lancetipton opened the pull request:
**Ticket**: [ZEN-450](https://jira.simpleviewtools.com/browse/ZEN-450)

> This PR should be tested in conjunction with these PRs
> [keg-cli](https://github.com/simpleviewinc/keg-cli/pull/49)
> [tap-events-force](https://github.com/simpleviewinc/tap-events-force/pull/127)

## Context
* Native / Mobile apps for keg-core have not been working for a while
* This needs to be fixed before we can start work for VisitApps
  * Current issues
    * Fonts don't load properly, and cause errors
    * Parts of Re-Theme can not be resolved within the Native App

### Updates

**Jest Tests**
* Added `--maxWorkers=50%` to the `test` command in the `package.json`
  * I read this [article](https://dev.to/vantanev/make-your-jest-tests-up-to-20-faster-by-changing-a-single-setting-i36) about speeding up tests runs.
  * Generally it seems the tests run much faster for `re-theme` and `tap-resolver`. I didn't really notice much for `keg-core` or `keg-components`
  * I haven't actually measured anything, but I figured it doesn't hurt so why not

**Keg-Core**
  * Updated how theme is initialized
    * Moved theme init to appInitAction call
    * May have to update taps as well, need to investigate
  * Fixed issue in app reducer
    * Cleaned up app.js UI
    * Added a fontLoader module
      * Allows loading custom fonts
    * Removed unused code
  * Added Retheme aliases for styleInjector && reStyle
    * Allows them to be resolved on Native apps
  * Added default font paths for custom fonts
    * Custom fonts directory must be within the assets directory
      * `tap/assets/fonts` => works
      * `tap/fonts` => does **NOT** work

**Keg-Components**
  * Fixed styles to work on all platforms
    * Removed use of overflow: 'none' - changed to 'hidden'
    * Removed default fontFamily in theme defaults
  * Fixed `accessibilityRole` in AppHeader
    * Set default to be `header`
    * Can only override when on web
  * Removed default fontFamily
    * Should be overridden from a taps theme

**Re-Theme**
  * Fixed issue with loading the correct platform
    * Added alias for RePlatform back
    * Without it, setRNPlatform had to be called
  * Added some jsDoc notes to the platform files
  * Added `auto-generated` index files
    * Removes the need for default index files that get copied on build

**Tap-Resolver**
  * Added auto-generated fonts based on asset file extension
  * fonts get split from other assets and an import file gets generated
    * Added defaults to constants for auto-generated asset files
    * Makes it easy to load custom fonts, see keg-core/actions/initAppAction for an example
      * Taps need the following to load a custom font
        * Add their fonts to a folder within the tap
        * Set the `tapFonts` folder path in the tap.js(on) at `keg.tapResolver.paths.tapFonts`
        * Call the `initAppAction` from keg-core within the tap 

### Tests

**Setup**
* Pull the changes from the `keg-cli` and `tap-events-force` repos
  * `keg cli && keg pr 49`
  * `keg evf && keg pr 127`
* Then pull these `keg-hub` changes `keg hub && keg pr 147`
* To fully test the native apps, you must have the Android emulator and IOS simulator setup
  * Setup is beyond this PR. If you need help, reach out and I'll help you get it setup

**Test 1**
* Pull this docker package and run it => `keg evf pack run evf:zen-450-fix-native-apps`
  * This the is the evf sessions component, but with the changes from keg-hub synced in
  * It also has the `Inter` font overwritten with `FiraSans-Regular`
    * This font change is only a temporary change within the tap to validate fonts are working
* When it starts, go to http://evf-zen-450-fix-native-apps.local.kegdev.xyz/ in your browser
  * Ensure the app starts as expected
  * Inspect a text item, and check it's text styles
    * Ensure the font-family is `FiraSans-Regular`
    * [Looks like this](http://snpy.in/LHQhFA)

**Test 2**
* Attach to the container with `keg d att package-evf-zen-450-fix-native-apps`
  * This will attach to the running docker container
  * Navigate to `src/assets/fonts`
    * Run the cmd `ls -la`
    * Ensure you see the `FiraSans-Regular` font along with the `Inter` fonts
  * Next, form the tap root navigate to `node_modules/keg-core/core/base/assets/fonts`
    * Run the cmd `cat index.js`
    * Ensure it prints the paths to the taps fonts, including `FiraSans-Regular`
    * [Looks like this](http://snpy.in/CYgVyW)
    * Ensure the paths to the fonts are pointing to the `tap-root/src/assets/fonts` folder

**Test 3**
> **Important**
> * This test requires either the Android emulator or IOS simulator
>   * Please ensure those are setup properly before running this test
> * Some of the styles are not perfect on the app
>  * Fixing this is outside the scope of this PR
>  * I did make a few changes, but there's more that are needed 

**Setup**
* These tests must be done on your host machine
* Which means you must copy the code from the keg-hub repos into the node_module of the tap
* Here's the steps to do that
  * Run `yarn install` the on keg-core => `keg core && yarn install`
  * Run `yarn install` the on re-theme => `keg retheme && yarn install`
    * Build retheme => `keg retheme && yarn build` 
    * Copy `re-theme` into the `node_modules` of `keg-core`
      * `keg core`
      * `rm -rf ./node_modules/@keg-hub/re-theme/build && cp -R ../re-theme/build ./node_modules/@keg-hub/re-theme/build`
  * Run `yarn install` the on keg-components => `keg components && yarn install`
    * Build keg-components => `keg components && yarn build` 
    * Copy `keg-components` into the `node_modules` of `keg-core`
      * `keg core`
      * `rm -rf ./node_modules/@keg-hub/keg-components/build && cp -R ../keg-components/build ./node_modules/@keg-hub/keg-components/build`
  * Run `yarn install` the on tap-resolver => `keg resolver && yarn install`
    * Copy `tap-resolver` into the `node_modules` of `keg-core`
      * `keg core`
      * `rm -rf ./node_modules/@keg-hub/tap-resolver && cp -R ../tap-resolver ./node_modules/@keg-hub/tap-resolver`
  * Run `yarn install` the on events-force tap => `keg evf && yarn install`
    * Copy `keg-core` into the the taps `node_modules` folder
      * `keg evf`
      * `rm -rf ./node_modules/keg-core && cp -R ~/keg-hub/repos/keg-core ./node_modules/keg-core`
* lol... and now you can test on mobile... it's that easy
  * One day we'll get this inside docker

**Test**
* Navigate to the events-force tap and run either ios or android or both
  * IOS => `keg evf && yarn ios`
  * Android => `keg evf && yarn android`
  * **NOTES** - If you have already run the app on IOS or Android
    * I recommend deleting it, and starting from scratch
    * Due to a number of updates, there can be issues with caching when the app starts
    * It better to just start fresh
      * To do this, just delete the app from the device, and run one of the above commands again
      * The app will be auto-installed onto the device
* Load the app in (IOS || Android) emulator
  * Take a screen capture of the app when it loads
* Next goto google fonts and download a custom font ( Anything other then `Inter` )
  * It must be either `.otf` || `.ttf` format
* After downloading;
  * Copy the font to the `src/assets/fonts` folder of the events-force tap
  * Then, double click the font and install it locally, which should open the Font-Book app
  * In Font-Book
    * Select the font from the list on the left and on the right you should see meta-data about the font
    * The first field should be `PostScript name`
    * Ensure the value of the `PostScript name` is the same name as the font file in `src/assets/fonts` folder of the events-force tap
      * [Looks like this](http://snpy.in/KrLcgu)
    * For example, with the current `Inter-Regular`
      * If the font file name was just `Inter`
        * **Which is it's name when downloading from google fonts**
      * The PostScript name is `Inter-Regular`
        * Because of this, the font will **NOT** load properly
      * I had to update the name from `Inter` file to be `Inter-Regular`
        * Then font worked as expected
* After ensuring the `PostScript name` and the custom font file name stored as `src/assets/fonts` are the same
  * Open `src/theme/typography.js`
    * Change lines 8 - 11, and enter the name of your custom font file
      ```
        const fonts = {
          headings: '<Custom-Font-Name (PostScript name)>',
          default: '<Custom-Font-Name (PostScript name)>',
        }
      ```
  * Kill the expo server, if it's already running, then start it agin
  * Next open (IOS || Android) emulator 
    * When the app loads, it should be using the custom font you loaded
    * Compare the current app with the screen capture from the previous step
      * [Looks like this](http://snpy.in/f59OXV)
        * Left Phone image the font is `FiraSans-Regular`
        * Right Phone image the font is the `default` for the device

mpcarolin commented on repos/tap-resolver/src/builders/buildAssets.js:
```diff
@@ -11,97 +36,214 @@ const tapConstants = require('../tap/tapConstants')
+ * @returns {string} - Passed in assetName if it does not contain spaces
+ */
+const validateNoSpace = assetName => {
+  if (!assetName.includes(' ')) return assetName
```
are other types of whitespace a problem with RN as well? Like tab characters? If so you might want to use a regex test with the general whitespace pattern `/\s+/g`

lancetipton replied:
This checks the name of a file on disk.
Any other `whitespace` characters are not allowed within file names, so I don't think it's needed.
Although it may be possible to hack an invalid character into a file name, I think the user would have bigger problems then if it can be loaded into the keg.

mpcarolin reviewed (changes requested):
iOS ran well, but I hit an issue with Android: it always crashes when trying to initially load. I didn't get any errors with this either. It just shows the UI for a split second then crashes.

I tried this on two separate simulators, a Pixel 3 on Android 10 and a Pixel 4 on Android 11, neither of which had run an expo app before, so they were fresh.

I also left a few comments on the code.

lancetipton commented on repos/re-theme/src/context/platform.native.js:
```diff
@@ -0,0 +1,37 @@
+ */
+const loadReactNativePlatform = () => {
+  const { Platform } = require('react-native')
+  RNPlatform = Platform
```
* The `require` cache we talked about was for node.
* This code should only ever run within react-native and  never on node, so it doesn't work the same way.
* Basically the `metro` bundler polyfills the global `require` method
  * I'm not 100% sure if the cache works the same way for at all within the `metro` environment
  * I do know that in a dev env, the `metro#require` actually makes a call to the backend server to load the content
  * This caches that call so it's only made once ( again, this is in a dev environment )
  * In production all require calls are run at built time and actually included with the bundle
    * At this point I have no idea how the `require` call  is resolved, so this just ensures is only run once

I think there would be some value in digging deeper into how `require` works within metro
  * There's this react-native thing call `ram-bundler`, that basically makes require calls dynamic
  * I have not tried it, but it may work in this case
  * It's definitely an area I plan to investigate more when I have the oppertunity

mpcarolin replied:
ah right, node. Alright makes sense.

lancetipton commented:
@mpcarolin - This is ready again.
* Pushed some fixes for accessibility that was causing issues. Added a new hook
* Switched to using expo-linear-gradient. The other repo was having issues

Android events-force tap no longer crashes.

mpcarolin commented on repos/keg-core/core/base/native/expo/fontLoader/fontLoader.js:
```diff
@@ -0,0 +1,22 @@
+
+  return !isObj(fontData)
+    ? false
+    : await Font.loadAsync(fontData)
```
Wow, that is fascinating and makes a lot of sense. The error wouldn't capture the function name if you just returned the promise b/c at that point I guess the promise isn't really tied to  function anymore in any way.

Interesting that they mention the tradeoff is a small performance hit with adding a task to the event loop. But I would prefer the more readable stack traces, so I'll start doing this too.

mpcarolin reviewed (approved):
Looks good
